cmd/reporting-operator/main.go

```diff
@@ -0,0 +1,112 @@
+package main
+
+import (
+	"context"
+	"errors"
+	"log"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/operate-first/curator-operator/internal"
+	"github.com/operate-first/curator-operator/pkg/operator"
+	"github.com/operate-first/curator-operator/pkg/signals"
+	"github.com/spf13/cobra"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+)
+
+const (
+	defaultServerPort = ":8080"
+)
+
+var (
+	reportLog = ctrl.Log.WithName("report")
+)
+
+var (
+	cfg operator.Config
+
+	rootCmd = &cobra.Command{
+		Use:   "reporting-operator",
+		Short: "",
+		RunE: func(cmd *cobra.Command, _ []string) error {
+			return cmd.Help()
+		},
+	}
+
+	startCmd = &cobra.Command{
+		Use:   "start",
+		Short: "Starts the http server",
+		Long:  "",
+		Run: func(cmd *cobra.Command, args []string) {
+			err := runReporting(context.Background(), reportLog)
+			if err != nil {
+				reportLog.Error(err, "error generating report")
+			}
+		},
+	}
+)
+
+func init() {
+	rootCmd.AddCommand(startCmd)
+	startCmd.Flags().StringVar(&cfg.Port, "port", defaultServerPort, "http server port")
+}
+
+func main() {
+	opts := zap.Options{
+		Development: true,
+	}
+	logger := zap.New(zap.UseFlagOptions(&opts))
+	ctrl.SetLogger(logger)
+
+	rootCmd.AddCommand(startCmd)
+	err := rootCmd.Execute()
+	if err != nil {
+		os.Exit(1)
+	}
+}
+
+func runReporting(ctx context.Context, logger logr.Logger) error {
+	var httpServerPort = cfg.Port
+	internal.NewHTTPServer(ctx, httpServerPort, logger)
+
+	server := http.Server{
+		Addr: httpServerPort,
+	}
+
+	logger.Info("starting http server on port" + httpServerPort)
+
+	go startServer(&server, logger)
+
+	stopCh, closeCh := signals.CreateChannel()
+	defer closeCh()
+	log.Println("system call notified:", <-stopCh)
+
+	err := shutdownServer(ctx, &server, logger)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func startServer(server *http.Server, logger logr.Logger) {
+	if err := server.ListenAndServe(); err != nil && !errors.Is(err, http.ErrServerClosed) {
+		logger.Error(err, "failed to listen on http server")
+	}
+	logger.Info("http server stopped gracefully")
+}
+
+func shutdownServer(ctx context.Context, server *http.Server, logger logr.Logger) error {
+	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+	defer cancel()
+
+	if err := server.Shutdown(ctx); err != nil {
+		logger.Error(err, "http server shutdown error:")
+		return err
+	}
+	logger.Info("http server shutdowned")
+
+	return nil
+}
```

Comment:
We're effectively trying to implement graceful shutdown for the server component but we're not catching any of the system calls that Kubernetes will fire off when rolling down a container. How do feel about the following implementation:

```go
func shutdownServer(timeout time.Duration, server *http.Server, logger logr.Logger) error {
	stop := make(chan os.Signal, 1)

	signal.Notify(stop, os.Interrupt, syscall.SIGTERM, syscall.SIGINT)
	sigReceived := <-stop
	logger.Info("received shutdown signal", "signal", sigReceived.String())

	ctx, cancel := context.WithTimeout(context.Background(), timeout)
	defer cancel()

	if err := server.Shutdown(ctx); err != nil {
		logger.Error(err, "failed to gracefull shutdown server")
		return err
	}
	return nil
}
```

I tacked that out inline so YMMV, but it _should_ catch the system calls we care about when in order to react and gracefully shutdown.

Reply:
I have added signals in another pkg i.e pkg/signals/signals.go. This implementation is similar to what we have already written.